### PR TITLE
Feature/te 295 add testmakro support

### PR DESCRIPTION
**Pull request by gunther-bachmann**

### CODE REVIEW

File: tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/scoping/integration/TclMacroCallIntegrationTest.xtend [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	@Inject TclJvmModelInferrer jvmModelInferrer // class under test
+	@Mock ITreeAppendable outputStub
+
+	override void setup() {
+		super.setup
+		when(outputStub.trace(any(EObject))).thenReturn(outputStub)
+		when(outputStub.append(any(CharSequence))).thenReturn(outputStub)
+	}
+
+	private def parseAmlModel(String aml) {
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/scoping/integration/TclMacroCallIntegrationTest.xtend]: This sound like a method we could also put to good use in the super class, couldn't we?

File: tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/scoping/integration/TclMacroCallIntegrationTest.xtend [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		val amlModel = amlParseHelper.parse(aml, resourceSet).assertNoSyntaxErrors
+		return amlModel
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/scoping/integration/TclMacroCallIntegrationTest.xtend]: Can be written in a single line, same for parseTmlModel (already done in parseTclModel).

File: tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/scoping/integration/TclMacroCallIntegrationTest.xtend [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+import static org.mockito.Matchers.*
+import static org.mockito.Mockito.*
+
+class TclMacroCallIntegrationTest extends AbstractTclGeneratorIntegrationTest {
+
+	@Inject TclJvmModelInferrer jvmModelInferrer // class under test
+	@Mock ITreeAppendable outputStub
+
+	override void setup() {
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/scoping/integration/TclMacroCallIntegrationTest.xtend]: I'd prefer a separate `@Before` method calle something like `configureMocks`

File: tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/contentassist/TclProposalProvider.xtend [hunk 4]
```diff
@@ -42,30 +27,4 @@ class TclProposalProvider extends AbstractTclProposalProvider {
 		acceptor.accept(createCompletionProposal('* ', '* test description', null, context))
 	}
 
-	override complete_TestStep(EObject model, RuleCall ruleCall, ContentAssistContext context,
-		ICompletionProposalAcceptor acceptor) {
-		acceptor.accept(createCompletionProposal('- ', '- test step', null, context))
-	}
-
-	override complete_StepContentElement(EObject model, RuleCall ruleCall, ContentAssistContext context,
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/contentassist/TclProposalProvider.xtend]: Why is this deleted?

File: tml/org.testeditor.tml.dsl/src/org/testeditor/tml/dsl/Tml.xtext [hunk 5]
```diff
@@ -1,10 +1,152 @@
...
+	'matches' | 'does' 'not' 'match' | 'is' 'not'? | 'template' | 'as' | 'Mask' | 'Component' | 'Macro' | 'implements' | 'package';
+
+StepContentVariable returns tsl::StepContentVariable:
+	value=STRING;
+
+StepContentElement hidden():
+	('<' value=ID '>') |
+	{StepContentElement} ('<>' | '<' WS+ '>');
+	
+StepContentDereferencedVariable:
```
Comment (franzbecker) [inline tml/org.testeditor.tml.dsl/src/org/testeditor/tml/dsl/Tml.xtext]: This should be called a `StepContentVariableReference`.

File: tml/org.testeditor.tml.dsl/src/org/testeditor/tml/dsl/Tml.xtext [hunk 6]
```diff
@@ -1,10 +1,152 @@
...
+	{StepContentDereferencedVariable}
+	'@' value=ID;
```
Comment (franzbecker) [inline tml/org.testeditor.tml.dsl/src/org/testeditor/tml/dsl/Tml.xtext]: This should be a direct reference on the variable.

File: tml/org.testeditor.tml.dsl.tests/src/org/testeditor/tml/dsl/tests/parser/TmlModelParserTest.xtend [hunk 7]
```diff
@@ -19,4 +19,30 @@ class TmlModelParserTest extends AbstractParserTest {
...
+		val input = '''
+			package com.example
+
+			# MyMacroCollection
+
+			template = "start with" ${startparam}
+			Component: MyComponent
+			- put @startparam into <other>
+
+			template = "use macro with" ${useparam}
```
Comment (franzbecker) [inline tml/org.testeditor.tml.dsl.tests/src/org/testeditor/tml/dsl/tests/parser/TmlModelParserTest.xtend]: A comment would be nice here that the macro from above is used.

File: tml/org.testeditor.tml.dsl/src/org/testeditor/tml/dsl/Tml.xtext [hunk 8]
```diff
@@ -1,10 +1,152 @@
...
 import "http://www.eclipse.org/emf/2002/Ecore" as ecore
 
 TmlModel:
 	'package' package=QualifiedName
 	(importSection=XImportSection)?
+	'#' name=ID
+	(macros+=Macro)*
 ;
+
+Macro:
```
Comment (franzbecker) [inline tml/org.testeditor.tml.dsl/src/org/testeditor/tml/dsl/Tml.xtext]: It would be nice to have something that wraps the macro, such as:
```
## myMacro 
   template = "start with" ${startparam}
    Component: MyComponent
      - put @startparam into <other>
```

File: tml/org.testeditor.tml.dsl.tests/src/org/testeditor/tml/dsl/validation/TmlVarUsageValidatorTest.xtend [hunk 9]
```diff
@@ -54,7 +55,7 @@ class TclVarUsageValidatorTest extends AbstractParserTest {
 		''')
 
 		// when
-		tclValidator.checkVariableUsageWithinAssertionExpressions(testStepContext)
+		tclValidator.checkVariableUsageWithinAssertionExpressions(testStepContext.assertInstanceOf(ComponentTestStepContext))
```
Comment (franzbecker) [inline tml/org.testeditor.tml.dsl.tests/src/org/testeditor/tml/dsl/validation/TmlVarUsageValidatorTest.xtend]: The `assertInstanceOf` check should not happen in the `// when` step. You can put it in the `// given` step, though.

File: tml/org.testeditor.tml.dsl.ui/src/org/testeditor/tml/dsl/ui/contentassist/TmlProposalProvider.xtend [hunk 10]
```diff
@@ -1,21 +1,68 @@
...
-* are made available under the terms of the Eclipse Public License v1.0
-* which accompanies this distribution, and is available at
-* http://www.eclipse.org/legal/epl-v10.html
-* 
-* Contributors:
-* Signal Iduna Corporation - initial API and implementation
-* akquinet AG
-* itemis AG
-*******************************************************************************/
+ * Copyright (c) 2012 - 2016 Signal Iduna Corporation and others.
```
Comment (franzbecker) [inline tml/org.testeditor.tml.dsl.ui/src/org/testeditor/tml/dsl/ui/contentassist/TmlProposalProvider.xtend]: This looks strange, what happened here?

File: tml/org.testeditor.tml.model/model/tml.xcore [hunk 11]
```diff
@@ -13,14 +13,81 @@
...
+	String name
+	contains Macro[0..*] macros
+}
+
+class Macro {
+	contains Template template
+	contains TestStepContext[0..*] contexts
+}
+
+class TestStepContext { }
```
Comment (franzbecker) [inline tml/org.testeditor.tml.model/model/tml.xcore]: class should be abstract

File: tml/org.testeditor.tml.model/model/tml.xcore [hunk 12]
```diff
@@ -13,14 +13,81 @@
...
+
+class TestStep {
+	contains StepContent[0..*] contents
+}
+
+class AssertionTestStep extends /*Component*/TestStep {
+	contains AssertionExpression expression
+}
+
+class AssertionExpression {}
```
Comment (franzbecker) [inline tml/org.testeditor.tml.model/model/tml.xcore]: class should be abstract

File: tml/org.testeditor.tml.model/model/tml.xcore [hunk 13]
```diff
@@ -13,14 +13,81 @@
...
+class MacroTestStepContext extends TestStepContext{
+	refers TmlModel macroModel
+	contains TestStep step
+}
+
+class TestStep {
+	contains StepContent[0..*] contents
+}
+
+class AssertionTestStep extends /*Component*/TestStep {
```
Comment (franzbecker) [inline tml/org.testeditor.tml.model/model/tml.xcore]: Why is the comment here?

File: tml/org.testeditor.tml.model/model/tml.xcore (hunk 11 above)
Comment (franzbecker) [inline tml/org.testeditor.tml.model/model/tml.xcore]: please reformat the file

File: tml/org.testeditor.tml.dsl/src/org/testeditor/tml/dsl/Tml.xtext (hunk 8 above)
Comment (gunther-bachmann) [inline tml/org.testeditor.tml.dsl/src/org/testeditor/tml/dsl/Tml.xtext]: corrected (correction does not invalidate this comment, since change applies to lines following...)

File: tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/contentassist/TclProposalProvider.xtend (hunk 4 above)
Comment (gunther-bachmann) [inline tcl/org.testeditor.tcl.dsl.ui/src/org/testeditor/tcl/dsl/ui/contentassist/TclProposalProvider.xtend]: Was moved to TmlProposalProvider, which is transitively inherited here
